clippy_lints/src/manual_clamp.rs

```diff
@@ -0,0 +1,733 @@
+        .or_else(|| to_do(&first_bin.flip(), &second_bin.flip()))
+        .or_else(|| to_do(&first_bin.flip(), second_bin))
+        .or_else(|| to_do(first_bin, &second_bin.flip()))
+}
+
+/// The clamp meta pattern is a pattern shared between many (but not all) patterns.
+/// In summary, this pattern consists of two if statements that meet many criteria,
+/// - binary operators that are one of [`>`, `<`, `>=`, `<=`].
+/// - Both binary statements must have a shared argument
+///     - Which can appear on the left or right side of either statement
+///     - The binary operators must define a finite range for the shared argument. To put this in
+///       the terms of Rust `std` library, the following ranges are acceptable
+///         - `Range`
+///         - `RangeInclusive`
+///       And all other range types are not accepted. For the purposes of `clamp` it's irrelevant
+///       whether the range is inclusive or not, the output is the same.
+/// - The result of each if statement must be equal to the argument unique to that if statement. The
+///   result can not be the shared argument in either case.
+fn is_clamp_meta_pattern<'tcx>(
+    cx: &LateContext<'tcx>,
+    first_bin: &BinaryOp<'tcx>,
+    second_bin: &BinaryOp<'tcx>,
+    first_expr: &'tcx Expr<'tcx>,
+    second_expr: &'tcx Expr<'tcx>,
+    // This parameters is exclusively for the match pattern.
+    // It exists because the variable bindings used in that pattern
+    // refer to the variable bound in the match arm, not the variable
+    // bound outside of it. Fortunately due to context we know this has to
+    // be the input variable, not the min or max.
+    input_hir_ids: Option<(HirId, HirId)>,
+) -> Option<InputMinMax<'tcx>> {
+    // First filter out any expressions with side effects
+    let exprs = [
+        first_bin.left,
+        first_bin.right,
+        second_bin.left,
+        second_bin.right,
+        first_expr,
+        second_expr,
+    ];
+    let clampability = TypeClampability::is_clampable(cx, cx.typeck_results().expr_ty(first_expr))?;
+    let is_float = clampability.is_float();
+    if exprs.iter().any(|e| simplify_kind(e).can_have_side_effects()) {
+        return None;
+    }
+    if !(is_ord_op(first_bin.op) && is_ord_op(second_bin.op)) {
+        return None;
+    }
+    do_with_flipping(first_bin, second_bin, |first_bin, second_bin| {
+        if !ord_op_same_direction(first_bin.op, second_bin.op) {
+            let min_is_first = match &first_bin.op {
+                BinOpKind::Ge | BinOpKind::Gt => eq_expr_value(cx, first_bin.left, first_expr),
+                BinOpKind::Le | BinOpKind::Lt => eq_expr_value(cx, first_bin.right, first_expr),
+                _ => unreachable!("prior check guarantees this is a comparison operator"),
+            };
+            let (min, max) = if min_is_first {
+                (first_expr, second_expr)
+            } else {
+                (second_expr, first_expr)
+            };
```

Comment:
I'd say make this an inner fn (or closure if nicer) rather than callback style so it's similar to the other functions in the lint

I find the `!ord_op_same_direction` + reversing `first_expr`/`second_expr` tough to get my head around what comparisons are permitted, it may be worth trying flipping the expression outside e.g.

```rust
/// for all of the `.flip()`y combos
check(first, second, input_hir_ids)
    .or_else(|| check(second, first, input_hir_ids.map(|(l, r)| (r, l))))
```

then inside checking just one orientation, such as `first_bin` being `Ge/Gt` and `second_bin` `Le/Lt`

Reply:
Just to be clear, you don't need to flip the `input_hir_ids`, that actually breaks the pattern detection. I did find some ways to simplify the flow of this code though.

Reply:
What I meant by that was swapping the arms outside `check`, something along the lines of

```rust
let cases = [(first_bin, second_bin), (first_bin.flip(), second_bin.flip), (first_bin.flip(), second_bin), (first_bin, second_bin.flip())];

cases.into_iter().find(|(first, second)| {
    // I think the exprs would need to be passed in as well so it's a bit longer now
    check(first, second, first_expr, second_expr, input_hir_ids)
        .or_else(|| check(second, first, second_expr, first_expr, input_hir_ids.map(|(l, r)| (r, l))))
})
```

And replacing

```rust
        if ord_op_same_direction(first_bin.op, second_bin.op) {
            return None;
        }
        let min_is_first = match &first_bin.op {
            BinOpKind::Ge | BinOpKind::Gt => eq_expr_value(cx, first_bin.left, first_expr),
            BinOpKind::Le | BinOpKind::Lt => eq_expr_value(cx, first_bin.right, first_expr),
            _ => unreachable!("prior check guarantees this is a comparison operator"),
        };
        let (min, max) = if min_is_first {
            (first_expr, second_expr)
        } else {
            (second_expr, first_expr)
        };
```

with

```rust
// or some equivalent
match (first_bin.op, second_bin.op) {
    (BinOpKind::Ge | BinOpKind::Gt, BinOpKind::Le | BinOpKind::Lt) => {},
    _ => return None
}
```

(I haven't tested this)

I'll leave the decision up to you

Reply:
Oh I see, that's a neat idea. Implemented!